proofs/tree.go

```diff
@@ -333,53 +310,138 @@ func getPropertyNameFromProtobufTag(tag string) (name string, err error) {
+	messageType    reflect.Type
+	messageValue   reflect.Value
+	excludedFields []string
+	salts          proto.Message
+	saltsValue     reflect.Value
+	leaves         LeafList
+	nodes          [][]byte
+	propOrder      []string
+}
+
+// generateLeafFromFieldIndex adds the LeafNode to LeafList for a field by it's
+// index in the Message struct.
+func (f *messageFlattener) generateLeafFromFieldIndex(index int) (err error) {
+	// Ignore fields starting with XXX_, those are protobuf internals
+	if strings.HasPrefix(f.messageType.Field(index).Name, "XXX_") {
+		return nil
+	}
 
-	for i := 0; i < v.NumField(); i++ {
-		value := v.Field(i).Interface()
-		tag := v.Type().Field(i).Tag.Get("protobuf")
+	tag := f.messageType.Field(index).Tag.Get("protobuf")
+	value := f.messageValue.Field(index).Interface()
+	prop, err := getPropertyNameFromProtobufTag(tag)
+	if err != nil {
+		return err
+	}
 
-		// Ignore fields starting with XXX_, those are protobuf internals
-		if strings.HasPrefix(v.Type().Field(i).Name, "XXX_") {
-			continue
+	// Check if the field has an exclude_from_tree option
+	for i := 0; i < len(f.excludedFields); i++ {
+		if f.excludedFields[i] == prop {
+			return
 		}
-		prop, err := getPropertyNameFromProtobufTag(tag)
+	}
+
+	salt := reflect.Indirect(f.saltsValue).FieldByName(f.messageType.Field(index).Name).Interface().([]byte)
+	leaf := LeafNode{
+		Property: prop,
+		Value:    value,
+		Salt:     salt,
+	}
+	f.leaves = append(f.leaves, leaf)
+	return nil
+}
+
+// sortLeaves by the property attribute and copies the properties and
+// concatenated byte values into the nodes
+func (f *messageFlattener) sortLeaves() (err error) {
+	sort.Sort(f.leaves)
+	f.nodes = make([][]byte, f.leaves.Len())
+	f.propOrder = make([]string, f.leaves.Len())
+
+	for i := 0; i < f.leaves.Len(); i++ {
+		f.nodes[i], err = ConcatNode(&f.leaves[i])
 		if err != nil {
-			return [][]byte{}, []string{}, err
+			return err
 		}
-		salt := reflect.Indirect(s).FieldByName(t.Field(i).Name).Interface().([]byte)
-		leaf := LeafNode{
-			Property: prop,
-			Value:    value,
-			Salt:     salt,
+		f.propOrder[i] = f.leaves[i].Property
+	}
+	return nil
+}
+
+// parseExtensions iterates over the prototype descripter to find fields that
+// should be excluded
+func (f *messageFlattener) parseExtensions() (err error) {
+	descriptorMessage, ok := f.message.(descriptor.Message)
+	if !ok {
+		return fmt.Errorf("message [%s] does not implement descriptor.Message", f.message)
+	}
+	_, messageDescriptor := descriptor.ForMessage(descriptorMessage)
+
+	for i := range messageDescriptor.Field {
+		fieldDescriptor := messageDescriptor.Field[i]
+		fV := reflect.ValueOf(fieldDescriptor).Elem()
+		fType := fV.Type()
+		fieldName := *fieldDescriptor.Name
+		for i := 0; i < fV.NumField(); i++ {
+			field := fV.Field(i)
+			if fType.Field(i).Name != "Options" {
+				continue
+			}
+			if proto.HasExtension(field.Interface().(proto.Message), proofspb.E_ExcludeFromTree) {
```

Comment:
If we already find that the field has the `ExcludeFromTreeExtension`, isn't it safe to assume that it should be excluded? (following on a previous comment)
So might not need to get the Extension value.